pkg/controller/nodeipam/ipam/multicidrset/cidr_set.go

```diff
@@ -0,0 +1,363 @@
+		return nil, ErrCIDRSetSubNetTooBig
+	}
+
+	// Register CidrSet metrics.
+	registerCidrsetMetrics()
+
+	maxCIDRs = 1 << uint32(subNetMaskSize-clusterMaskSize)
+	return &CidrSet{
+		clusterCIDR:      clusterCIDR,
+		nodeMask:         net.CIDRMask(subNetMaskSize, bits),
+		clusterMaskSize:  clusterMaskSize,
+		maxCIDRs:         maxCIDRs,
+		nodeMaskSize:     subNetMaskSize,
+		label:            clusterCIDR.String(),
+		AllocatedCIDRMap: make(map[string]bool, 0),
+	}, nil
+}
+
+func (s *CidrSet) indexToCIDRBlock(index int) *net.IPNet {
+	var ip []byte
+	switch /*v4 or v6*/ {
+	case s.clusterCIDR.IP.To4() != nil:
+		{
+			j := uint32(index) << uint32(32-s.nodeMaskSize)
+			ipInt := (binary.BigEndian.Uint32(s.clusterCIDR.IP)) | j
+			ip = make([]byte, net.IPv4len)
+			binary.BigEndian.PutUint32(ip, ipInt)
+		}
+	case s.clusterCIDR.IP.To16() != nil:
+		{
+			// leftClusterIP      |     rightClusterIP
+			// 2001:0DB8:1234:0000:0000:0000:0000:0000
+			const v6NBits = 128
+			const halfV6NBits = v6NBits / 2
+			leftClusterIP := binary.BigEndian.Uint64(s.clusterCIDR.IP[:halfIPv6Len])
+			rightClusterIP := binary.BigEndian.Uint64(s.clusterCIDR.IP[halfIPv6Len:])
+
+			ip = make([]byte, net.IPv6len)
+
+			if s.nodeMaskSize <= halfV6NBits {
+				// We only care about left side IP.
+				leftClusterIP |= uint64(index) << uint(halfV6NBits-s.nodeMaskSize)
+			} else {
+				if s.clusterMaskSize < halfV6NBits {
+					// see how many bits are needed to reach the left side.
+					btl := uint(s.nodeMaskSize - halfV6NBits)
+					indexMaxBit := uint(64 - bits.LeadingZeros64(uint64(index)))
+					if indexMaxBit > btl {
+						leftClusterIP |= uint64(index) >> btl
+					}
+				}
+				// the right side will be calculated the same way either the
+				// subNetMaskSize affects both left and right sides.
+				rightClusterIP |= uint64(index) << uint(v6NBits-s.nodeMaskSize)
+			}
+			binary.BigEndian.PutUint64(ip[:halfIPv6Len], leftClusterIP)
+			binary.BigEndian.PutUint64(ip[halfIPv6Len:], rightClusterIP)
+		}
+	}
+	return &net.IPNet{
+		IP:   ip,
+		Mask: s.nodeMask,
+	}
+}
+
+// NextCandidate returns the next candidate for the current cidrSet.
+// Returns nil if the candidate is already allocated.
+func (s *CidrSet) NextCandidate() (*net.IPNet, int, error) {
+	s.Lock()
+	defer s.Unlock()
+
+	if s.allocatedCIDRs == s.maxCIDRs {
+		return nil, 0, ErrCIDRRangeNoCIDRsRemaining
+	}
+
+	var i int
+	var nextCandidateCIDR *net.IPNet
+
+	candidate := s.nextCandidate
+	for i = 0; i < s.maxCIDRs; i++ {
+		nextCandidateCIDR = s.indexToCIDRBlock(candidate)
+		// Check if the nextCandidate is not already allocated.
+		if _, ok := s.AllocatedCIDRMap[nextCandidateCIDR.String()]; !ok {
+			break
+		}
+		candidate = (candidate + 1) % s.maxCIDRs
+	}
+
+	s.nextCandidate = (candidate + 1) % s.maxCIDRs
+
+	return nextCandidateCIDR, i, nil
+}
+
+func (s *CidrSet) getBeginningAndEndIndices(cidr *net.IPNet) (begin, end int, err error) {
```

Comment:
don't use named return unless function is trivial (it is very easy to accidentally shadow return var

Reply:
Ack, modified